### PR TITLE
Allow plugins to install dependencies during build

### PR DESCRIPTION
**Pull request by jbeezley**
The purpose of this PR is actually to make the build process a little bit more flexible for plugins needing to do special things.  I ended up splitting the different build components into separate files so that they are a little less jumbled together.  Most of the content remains the same with the exception of the plugin related targets.

Major changes:
* `init` and `default` targets can be modified by plugins via the [`config`](https://github.com/girder/girder/blob/a4f9a89f619fcd4875bb5870dd247af584b41046/Gruntfile.js#L80) object
* There now exists [methods](https://github.com/girder/girder/blob/a4f9a89f619fcd4875bb5870dd247af584b41046/docs/plugin-development.rst#installing-custom-dependencies-from-npm) for plugins to install dependencies both within
their own `node_modules` directory and within girder's `node_modules` directory.

### CODE REVIEW

File: docs/plugin-development.rst [hunk 1]
```diff
@@ -435,6 +435,44 @@ that can be used to import content:
   plugin that doesn't fall into one of the above categories can be placed here,
   such as static images, fonts, or third-party static libraries.
 
+Installing custom dependencies from npm
+***************************************
+
+There are two types of node dependencies you may need to install for your plugin.
+Each type need to be installed differently due to how node manages external packages.
```
Comment (zachmullen) [inline docs/plugin-development.rst]: *needs

File: grunt_tasks/build.js [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+                tasks: ['stylus:core']
+            },
+            js_core: {
+                files: ['clients/web/src/**/*.js'],
+                tasks: ['uglify:app']
+           },
+            jade_core: {
+                files: ['clients/web/src/templates/**/*.jade'],
+                tasks: ['build-js']
+            }
```
Comment (danlamanna) [inline grunt_tasks/build.js]: This task doesn't exist anymore.

File: grunt_tasks/plugin.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+/**
+ * Define tasks related to loading, configuring, and building plugins.
+ */
+module.exports = function (grunt) {
+
+    var _ = require('underscore');
+    var fs = require('fs');
+    var path = require('path');
+    var defaultTasks = grunt.config.get('defaultTasks');
```
Comment (danlamanna) [inline grunt_tasks/plugin.js]: Is this being used anymore?

Comment (jbeezley) [inline grunt_tasks/plugin.js]: No, good eye.

**Comment by danlamanna** [comment]
(This is directly related to something I was trying to work on (#1046)), which is to make any stylus tasks dependent on ```shell:readServerConfig```.

This looks cool, I like being able to specify dependencies in a way other than ```grunt.task.requires``` which doesn't actually run the required tasks. I'll play around with it later when looking at my use case.

**Comment by jbeezley** [comment]
I attempted to add javascript style tests for the build scripts, but I'm hitting a strange cmake error.

```
$ ctest -R grunt -VV
UpdateCTestConfiguration  from :/Users/jbeezley/kwgit/girder/_build/DartConfiguration.tcl
Parse Config file:/Users/jbeezley/kwgit/girder/_build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/Users/jbeezley/kwgit/girder/_build/DartConfiguration.tcl
Parse Config file:/Users/jbeezley/kwgit/girder/_build/DartConfiguration.tcl
Test project /Users/jbeezley/kwgit/girder/_build
Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 27
    Start 27: jshint_grunt_main

27: Test command: /Users/jbeezley/.nvm/versions/io.js/v2.3.1/bin/jshint "--config" "/Users/jbeezley/kwgit/girder/_build/tests/grunt_main_jshint.cfg" "/Users/jbeezley/kwgit/girder/Gruntfile.js"
27: Test timeout computed to be: 1500
1/4 Test #27: jshint_grunt_main ................***Not Run   0.00 sec
test 28
    Start 28: jsstyle_grunt_main

28: Test command: /Users/jbeezley/.nvm/versions/io.js/v2.3.1/bin/jscs "--verbose" "--config" "/Users/jbeezley/kwgit/girder/_build/tests/grunt_main_jsstyle.cfg" "/Users/jbeezley/kwgit/girder/Gruntfile.js"
28: Test timeout computed to be: 1500
2/4 Test #28: jsstyle_grunt_main ...............***Not Run   0.00 sec
test 29
    Start 29: jshint_grunt_modules

29: Test command: /Users/jbeezley/.nvm/versions/io.js/v2.3.1/bin/jshint "--config" "/Users/jbeezley/kwgit/girder/_build/tests/grunt_modules_jshint.cfg" "/Users/jbeezley/kwgit/girder/grunt_tasks"
29: Test timeout computed to be: 1500
3/4 Test #29: jshint_grunt_modules .............   Passed    0.38 sec
test 30
    Start 30: jsstyle_grunt_modules

30: Test command: /Users/jbeezley/.nvm/versions/io.js/v2.3.1/bin/jscs "--verbose" "--config" "/Users/jbeezley/kwgit/girder/_build/tests/grunt_modules_jsstyle.cfg" "/Users/jbeezley/kwgit/girder/grunt_tasks"
30: Test timeout computed to be: 1500
30: No code style errors found.
4/4 Test #30: jsstyle_grunt_modules ............   Passed    0.33 sec

The following tests passed:
	jshint_grunt_modules
	jsstyle_grunt_modules

50% tests passed, 2 tests failed out of 4

Total Test time (real) =   0.72 sec

The following tests FAILED:
	 27 - jshint_grunt_main (BAD_COMMAND)
	 28 - jsstyle_grunt_main (BAD_COMMAND)
Errors while running CTest
```

Running the command works fine.  I guess it has to do with providing a single file as a target rather than a directory, but I can't figure out where the `BAD_COMMAND` error is coming from.

**Comment by zachmullen** [comment]
BAD_COMMAND probably means the executable (the first argument) is not a valid executable, maybe it doesn't exist.

**Comment by jbeezley** [comment]
But it does.  Also the same commands work in the other two tests I added.

**Comment by zachmullen** [comment]
If you change the value of the CMake variable `JSHINT_EXECUTABLE` to point to the local one under `node_modules/jshint/bin`, does it work then?

**Comment by jbeezley** [comment]
No difference.  Also, the same error occurs on travis.

```
$ ctest -R grunt -VV
UpdateCTestConfiguration  from :/Users/jbeezley/kwgit/girder/_build/DartConfiguration.tcl
Parse Config file:/Users/jbeezley/kwgit/girder/_build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/Users/jbeezley/kwgit/girder/_build/DartConfiguration.tcl
Parse Config file:/Users/jbeezley/kwgit/girder/_build/DartConfiguration.tcl
Test project /Users/jbeezley/kwgit/girder/_build
Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 27
    Start 27: jshint_grunt_main

27: Test command: /Users/jbeezley/kwgit/girder/node_modules/.bin/jshint "--config" "/Users/jbeezley/kwgit/girder/_build/tests/grunt_main_jshint.cfg" "/Users/jbeezley/kwgit/girder/Gruntfile.js"
27: Test timeout computed to be: 1500
1/4 Test #27: jshint_grunt_main ................***Not Run   0.00 sec
test 28
    Start 28: jsstyle_grunt_main

28: Test command: /Users/jbeezley/kwgit/girder/node_modules/.bin/jscs "--verbose" "--config" "/Users/jbeezley/kwgit/girder/_build/tests/grunt_main_jsstyle.cfg" "/Users/jbeezley/kwgit/girder/Gruntfile.js"
28: Test timeout computed to be: 1500
2/4 Test #28: jsstyle_grunt_main ...............***Not Run   0.00 sec
test 29
    Start 29: jshint_grunt_modules

29: Test command: /Users/jbeezley/kwgit/girder/node_modules/.bin/jshint "--config" "/Users/jbeezley/kwgit/girder/_build/tests/grunt_modules_jshint.cfg" "/Users/jbeezley/kwgit/girder/grunt_tasks"
29: Test timeout computed to be: 1500
3/4 Test #29: jshint_grunt_modules .............   Passed    0.35 sec
test 30
    Start 30: jsstyle_grunt_modules

30: Test command: /Users/jbeezley/kwgit/girder/node_modules/.bin/jscs "--verbose" "--config" "/Users/jbeezley/kwgit/girder/_build/tests/grunt_modules_jsstyle.cfg" "/Users/jbeezley/kwgit/girder/grunt_tasks"
30: Test timeout computed to be: 1500
30: No code style errors found.
4/4 Test #30: jsstyle_grunt_modules ............   Passed    0.30 sec

The following tests passed:
	jshint_grunt_modules
	jsstyle_grunt_modules

50% tests passed, 2 tests failed out of 4

Total Test time (real) =   0.66 sec

The following tests FAILED:
	 27 - jshint_grunt_main (BAD_COMMAND)
	 28 - jsstyle_grunt_main (BAD_COMMAND)
Errors while running CTest
```

**Comment by zachmullen** [comment]
Really strange...

**Comment by jbeezley** [comment]
I see, it is the working directory option, I'll work out a fix.

```
  add_test(
    NAME "jshint_${name}"
    WORKING_DIRECTORY ${input}
    COMMAND "${JSHINT_EXECUTABLE}" --config "${jshint_config}" "${input}"
  )
```

**Comment by jbeezley** [comment]
Should be good now.  PTAL

**Comment by jbeezley** [comment]
@danlamanna If you need a task to depend on another, you should assert it via [`grunt.task.requires`](http://gruntjs.com/api/grunt.task#grunt.task.requires) or [`grunt.config.requires`](http://gruntjs.com/api/grunt.config#grunt.config.requires).  I added a few of these in this PR.

You can do things like queuing tasks via [`grunt.task.run`](http://gruntjs.com/api/grunt.task#grunt.task.run) to make sure it run automatically which I found to get a little hairy if you try doing it too much.

**Comment by zachmullen** [comment]
This is great, and at some point soon I'd like to also expose bower as a mechanism for installing runtime dependencies for plugins.

LGTM.

**Pull request merged**